### PR TITLE
update Neuron SDK in LifeCycleScript

### PR DESCRIPTION
**Pull request by KeitaW**
The HyperPod AMI release process does not sync with Neuron SDK release schedule, this PR is proposing to update them in LifeCycleScript.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    # Install git
+    sudo apt-get install git -y
+
+    # Remove preinstalled packages and Install Neuron Driver and Runtime
+    sudo apt-get remove aws-neuron-dkms -y
+    sudo apt-get remove aws-neuronx-dkms -y
+    sudo apt-get remove aws-neuronx-oci-hook -y
+    sudo apt-get remove aws-neuronx-runtime-lib -y
+    sudo apt-get remove aws-neuronx-collectives -y
+    sudo apt-get install aws-neuronx-dkms=2.17.17.0 -y
```
Comment (KeitaW) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh:24]:
```suggestion
    sudo apt-get install aws-neuronx-dkms=2.* -y
```

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    sudo apt-get install aws-neuronx-oci-hook=2.4.4.0 -y
+    sudo apt-get install aws-neuronx-runtime-lib=2.21.41.0* -y
+    sudo apt-get install aws-neuronx-collectives=2.21.46.0* -y
```
Comment (KeitaW) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh:27]:
```suggestion
    sudo apt-get install aws-neuronx-collectives=2.* -y
```

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    # Remove preinstalled packages and Install Neuron Driver and Runtime
+    sudo apt-get remove aws-neuron-dkms -y
+    sudo apt-get remove aws-neuronx-dkms -y
+    sudo apt-get remove aws-neuronx-oci-hook -y
+    sudo apt-get remove aws-neuronx-runtime-lib -y
+    sudo apt-get remove aws-neuronx-collectives -y
+    sudo apt-get install aws-neuronx-dkms=2.17.17.0 -y
+    sudo apt-get install aws-neuronx-oci-hook=2.4.4.0 -y
+    sudo apt-get install aws-neuronx-runtime-lib=2.21.41.0* -y
```
Comment (KeitaW) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh:26]:
```suggestion
    sudo apt-get install aws-neuronx-runtime-lib=2.* -y
```

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    sudo apt-get install aws-neuronx-collectives=2.21.46.0* -y
+
+    # Remove pre-installed package and Install Neuron Tools
+    sudo apt-get remove aws-neuron-tools  -y
+    sudo apt-get remove aws-neuronx-tools  -y
+    sudo apt-get install aws-neuronx-tools=2.18.3.0 -y
```
Comment (KeitaW) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh:32]:
```suggestion
    sudo apt-get install aws-neuronx-tools=2.* -y
```

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh [hunk 5]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+set -euxo pipefail
+if neuron-top -v &> /dev/null
+then 
+    echo "Update to Neuron SDK Release 2.19.0"
```
Comment (KeitaW) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh:5]:
```suggestion
    echo "Update to the latest Neuron SDK"
```

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh [hunk 6]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+set -euxo pipefail
+if neuron-top -v &> /dev/null
+then 
+    echo "Update to the latest Neuron SDK"
+    # Configure Linux for Neuron repository updates
+    . /etc/os-release
+
+    sudo echo "deb https://apt.repos.neuron.amazonaws.com ${VERSION_CODENAME} main" | sudo tee /etc/apt/sources.list.d/neuron.list
```
Comment (shimomut) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh:9]: Where the "VERSION_CODENAME" variable is defined?
I assume this variable is to explicitly specify the SDK version to install, and it is a good idea, but cannot find it in this shell script itself.

**Review by shimomut — COMMENTED** [review]
Let's review this idea itself with the dev team as well. Also need to check if there is any impact on hardware health check behavior, etc.

**Review by nghtm — COMMENTED** [review]
Sharing the results of my test, where I ran the `update_neuron_sdk.sh` script directly on a HyperPod trn1-32.xlarge worker node. 

@KeitaW , can you please confirm whether this is intended behavior? I was expecting to see 2.19*

```
$ neuron-ls --version
neuron-ls 2.17.1.0%kaena-tools/2.17@685ebdc built on 2024-01-26T19:01:50Z

$ bash update_neuron_sdk.sh 
.... 
(output logs attached)
...

$ neuron-ls --version
neuron-ls 2.18.3.0%kaena-tools/2.18@e7386d8 built on 2024-05-16T02:23:38Z
```
[update-neuron-output.txt](https://github.com/user-attachments/files/16444479/update-neuron-output.txt)

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh (hunk 6 above)
Comment (nghtm) [reply 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh:9]: without specifying VERSION_CODENAME, the script updated me to 2.18.3 


(see here: https://github.com/aws-samples/awsome-distributed-training/pull/390#pullrequestreview-2210596156)

Comment (KeitaW) [reply 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh:9]:
```
$ cat /etc/os-release 
NAME="Ubuntu"
VERSION="20.04.6 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.6 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```
The `VERSION_CODENAME` is indicating ubuntu version and defined in `/etc/os-release`.

**Comment by KeitaW** [comment]
> Sharing the results of my test, where I ran the `update_neuron_sdk.sh` script directly on a HyperPod trn1-32.xlarge worker node.
> 
> @KeitaW , can you please confirm whether this is intended behavior? I was expecting to see 2.19*
> 
> ```
> $ neuron-ls --version
> neuron-ls 2.17.1.0%kaena-tools/2.17@685ebdc built on 2024-01-26T19:01:50Z
> 
> $ bash update_neuron_sdk.sh 
> .... 
> (output logs attached)
> ...
> 
> $ neuron-ls --version
> neuron-ls 2.18.3.0%kaena-tools/2.18@e7386d8 built on 2024-05-16T02:23:38Z
> ```
> 
> [update-neuron-output.txt](https://github.com/user-attachments/files/16444479/update-neuron-output.txt)

Thank you for the comment @nghtm .  The `neuron-ls` or any other tool version is not sync with "Neuron SDK" version (actually it is thighed with "System Tools" version).   According to the [release note](https://awsdocs-neuron.readthedocs-hosted.com/en/v2.19.1/release-notes/index.html#latest-neuron-release-artifacts), Neuron SDK 2.19.1 (latest as of today) includes the following packages.


```
List of packages in Neuron 2.19.1 (picked up few relevant to this discussion):
Component                           Package                                           
Collective Communication Library    aws-neuronx-collectives-2.21.46.0 
Driver                              aws-neuronx-dkms-2.17.17.0 
OCI                                 aws-neuronx-oci-hook-2.4.4.0
Runtime Library                     aws-neuronx-runtime-lib-2.21.41.0 
System Tools                        aws-neuronx-tools-2.18.3.0
```

and you can see if those are installed with the command as follows:

```
ubuntu@ip-10-1-10-50:~$ srun -N1 dpkg -l | grep neuron 
ii  aws-neuronx-collectives                2.21.46.0-69b77134b                   amd64        neuron_ccom built using CMake
ii  aws-neuronx-dkms                       2.17.17.0                             amd64        aws-neuronx driver in DKMS format.
ii  aws-neuronx-oci-hook                   2.4.4.0                               amd64        neuron_oci_hook built using CMake
ii  aws-neuronx-runtime-lib                2.21.41.0-fb1705f5f                   amd64        neuron_runtime built using CMake
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
```


and you can

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh [hunk 7]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
```
Comment (mhuguesaws) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh:1]: Please convert this to Ansible in order to do live update on compute nodes.

**Review by mhuguesaws — COMMENTED** [review]
left comments

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh (hunk 7 above)
Comment (KeitaW) [reply 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh:1]: Not sure SMHP works with ansible. @shimomut  could you confirm?

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh (hunk 6 above)
Comment (shimomut) [reply 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh:9]: @KeitaW , thank you for the answer about where VERSION_CODENAME is defined.

**Comment by shimomut** [comment]
Does this script try to install the latest version of the Neuron SDK always?
Let's make sure we are considering version consistency between nodes in a cluster.
e.g., when instance replacement happens 3 months after the cluster creation, do all the nodes use same SDK version?

**Comment by KeitaW** [comment]
> Does this script try to install the latest version of the Neuron SDK always? Let's make sure we are considering version consistency between nodes in a cluster. e.g., when instance replacement happens 3 months after the cluster creation, do all the nodes use same SDK version?

This is a good point. Currently, the script installs the latest SDK version available. At present, this version is 2.19.1, but it could be 2.20 or higher in three months. The initial approach I considered involved fixing the SDK version in the lifecycle script and periodically updating it in this repository.

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh (hunk 7 above)
Comment (shimomut) [reply 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/update_neuron_sdk.sh:1]: I am also not sure about Ansible on HyperPod, but in my option, this script can be in this shell script format as same as other parts of the lifecycle configuration script. HyperPod's default lifecycle script is written in shell scripts and python scripts, and we can follow the current style. Open to other opinions and discussion.

**Comment by nghtm** [comment]
I suggest going with commit [1e92461e44bfd388908a2304e88434d5d85e56b3](https://github.com/aws-samples/awsome-distributed-training/pull/390/commits/1e92461e44bfd388908a2304e88434d5d85e56b3) which contains hard coded SDK versions due to [this conversation](https://github.com/aws-samples/awsome-distributed-training/pull/390#issuecomment-2264153196). We can always provide instructions for customers to run the LCS via srun to get latest versions if required.

One final thought, should this be a configurable via config.py like the observability sollution, or do we make this LCS default? I think configurable (opt-in) is best.

**Comment by nghtm** [comment]
Tested 16 node cluster deployment with Commit 16fdad6e82dc12a5d68096bb36e29f9f725fec38 enabled (latest commit for this PR)...

Set  `enable_observability = True` in config.py.

Confirmed successful cluster deployment and version updates on the nodes:

```
$ srun -N1 dpkg -l | grep neuron 
ii  aws-neuronx-collectives                2.21.46.0-69b77134b                   amd64        neuron_ccom built using CMake
ii  aws-neuronx-dkms                       2.17.17.0                             amd64        aws-neuronx driver in DKMS format.
ii  aws-neuronx-runtime-lib                2.21.41.0-fb1705f5f                   amd64        neuron_runtime built using CMake
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools

$ sinfo
PARTITION        AVAIL  TIMELIMIT  NODES  STATE NODELIST
dev*                up   infinite     16   idle ip-10-1-0-44,ip-10-1-1-31,ip-10-1-3-177,ip-10-1-6-80,ip-10-1-10-78,ip-10-1-50-140,ip-10-1-51-17,ip-10-1-56-188,ip-10-1-83-33,ip-10-1-87-[23,54],ip-10-1-90-176,ip-10-1-94-90,ip-10-1-107-36,ip-10-1-108-90,ip-10-1-116-148
ml.trn1.32xlarge    up   infinite     16   idle ip-10-1-0-44,ip-10-1-1-31,ip-10-1-3-177,ip-10-1-6-80,ip-10-1-10-78,ip-10-1-50-140,ip-10-1-51-17,ip-10-1-56-188,ip-10-1-83-33,ip-10-1-87-[23,54],ip-10-1-90-176,ip-10-1-94-90,ip-10-1-107-36,ip-10-1-108-90,ip-10-1-116-148

$ srun -N16 dpkg -l | grep aws-neuronx-tools
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
ii  aws-neuronx-tools                      2.18.3.0                              amd64        Neuron profile and debug tools
```

Approving PR

**Review by nghtm — APPROVED** [review]
Approving after successful testing of latest comit:

https://github.com/aws-samples/awsome-distributed-training/pull/390/commits/ba8c4f6bb33b4e8c16b07cd360be3250ab0bfa87

**Pull request merged**